server/app/views/components/QuestionBank.java

```diff
@@ -219,50 +232,65 @@ private DivTag renderQuestionDefinition(QuestionDefinition definition) {
    *
    * <ul>
-   *   <li>If there is at least one question in the current block, all single-block questions are
-   *       filtered.
-   *   <li>If there is a single block question in the current block, all questions are filtered.
+   *   <li>If there is at least one question in the block, all single-block questions are filtered.
+   *   <li>If there is a single block question in the block, all questions are filtered.
    *   <li>If this is a repeated block, only the appropriate repeated questions are showed.
    *   <li>Questions already in the program are filtered.
    * </ul>
    */
-  private Stream<QuestionDefinition> filterQuestions() {
-    if (containsSingleBlockQuestion()) {
-      return Stream.empty();
+  private boolean canAddQuestionToBlock(QuestionDefinition question, BlockDefinition block) {
+    if (containsSingleBlockQuestion(block)) {
+      return false;
     }
-
-    Predicate<QuestionDefinition> filter =
-        params.blockDefinition().getQuestionCount() > 0
-            ? this::nonEmptyBlockFilter
-            : this::questionFilter;
-    return params.questions().stream().filter(filter);
-  }
-
-  /** If a block already contains a single-block question, no more questions can be added. */
-  private boolean containsSingleBlockQuestion() {
-    return params.blockDefinition().isEnumerator() || params.blockDefinition().isFileUpload();
+    if (block.getQuestionCount() > 0 && singleBlockQuestion(question)) {
+      return false;
+    }
+    if (params.program().hasQuestion(question)) {
+      return false;
+    }
+    // if question is not related to an enumerator - its enumerator id will be
+    // empty Optional, same as getEnumberatorQuestionId for block.
+    return question.getEnumeratorId().equals(getEnumeratorQuestionId(block));
   }
 
   /**
-   * An block can add questions with the appropriate enumerator id that aren't already in this
-   * program.
+   * Get list of blocks to which the given question can be added. See {@code
+   * canAddQuestionToBlock()} for conditions.
    */
-  private boolean questionFilter(QuestionDefinition questionDefinition) {
-    return questionDefinition.getEnumeratorId().equals(getEnumeratorQuestionId())
-        && !params.program().hasQuestion(questionDefinition);
+  private ImmutableSet<Long> getBlocksThatQuestionEligibleFor(QuestionDefinition question) {
+    return params.program().blockDefinitions().stream()
+        .filter(b -> this.canAddQuestionToBlock(question, b))
+        .map(BlockDefinition::id)
+        .collect(ImmutableSet.toImmutableSet());
   }
 
-  /**
-   * A non-empty block cannot add single-block questions, in addition to {@link
-   * QuestionBank#questionFilter}.
-   */
-  private boolean nonEmptyBlockFilter(QuestionDefinition questionDefinition) {
-    return !singleBlockQuestion(questionDefinition) && questionFilter(questionDefinition);
+  /** If a block already contains a single-block question, no more questions can be added. */
```

Comment:
The comment is more relevant to the caller not this method which doesn't enforce the comment's statement.